### PR TITLE
Add base camp construction markers

### LINKED ISSUE #71016: Introduce a construction site indication for base camp/expansion construction in progress
### Is your feature request related to a problem? Please describe.

The current base camp construction process does not tell the player where the construction is actually taking place, so the player has little indication of where not to place stuff in the mean time, resulting in the risk of walls obliterating valuable items.

### Solution you would like.

When the construction work is started, the construction area should be indicated in some fashion.
When the construction is finished the indication should be removed.
It's probably acceptable if indications for aborted constructions remain (aborted jobs just removes the job, so it's hard to add any cleanup to it, and the construction will probably be restarted some time in the future).

### Describe alternatives you have considered.

We've been fumbling in the dark thus far, and can most probably continue to do so without any severe lasting harm.

### Additional context

The approach I've thought of (which doesn't mean it's the only one, or the best one) is to use a process along these lines:
- Process the construction recipe to find the tiles where things are modified (i.e. ignoring the tiles that do nothing) and transfer these to a construction order for the whole world tile with most tiles being no op, and the modified ones containing the indicator. The processing should carry over rotation/mirroring.
- My current thought is that a field might be a good indicator, since you can have several fields on a single tile, so it shouldn't clash with anything.
- When the construction order is finalized, the process above is executed and thus the indicators are placed.
- When the construction is finished, the same logic as above is performed to generate an indicator removal order (quite probably the same operation is used to generate this, with a flag to indicate whether placement or removal is wanted).
- Once the removal is generated, it's applied to the world tile, followed by the actual construction finalization.

One issue is that I'd want the removal step to remove only the indicator, and not anything else of the same type on the tile. Thus, if fields were to be used, there would be a need to have an order that erases only a specific field (if present) on the tile, not all fields.

### PR DESCRIPTION
**Pull request by PatrikLundell**
#### Summary

None

#### Purpose of change

Fix #71016, as per the title

#### Describe the solution

Detect which tiles are changed by the application of a construction order by applying them to a fake map and then apply the construction site field to the corresponding tiles of the real map when initiating construction and doing the corresponding action to remove the construction field site after construction.

The logic was mostly stolen from an operation that counted number of changed terrain and furniture tiles.

This required the introduction of a new field, with the "looks like" selected to hopefully not cause any confusion.

It can be noted that the author can not be trusted to draw a stick figure without supervision, so tile support will have to be provided by competent people.

Apply the solution above to salt water pipes and remote fortifications.

Add new map operation to get rid of fields directly, rather than waiting for maintenance updates that don't happen.

#### Describe alternatives you've considered

- Remove mid 0.F backward compatibility code from remote fortifications. Deemed a little early.
- Change remote fortification implementation to use rotation/mirroring. Out of scope, provides no game play benefit, and would have insignificant effect on the code volume.

#### Testing

- Order the construction of a livestock version 2 chicken coop (it's subjected to both rotation and mirroring, if I remember correctly).
- Order the construction of a remote defense line.
- Go out and check that it's marked.
- 
Started the defense line project:
Screenshot (313)

- Expand camp with a Workshop version 2.
- Order the construction of the start of a salt water pipe.
- Go outside and verify the fields appear on the ground for the chicken coop.

Started the chicken coop project:
Screenshot (314)

- Continue and check on the salt water pipe project.

Started the salt water pipe project:
Screenshot (315)

- Return, debug advance time, and finish the three project steps.
- Go out and check on the progress.

Finished the chicken coop:
Screenshot (316)

Finished the trench part of the defense line:
Screenshot (317)

Finished the first part of the salt water pipe project (well in the swamp):
Screenshot (318)

- Return back and order the construction of spikes in the defense line trenches, as well as a set of salt water pipes.
- Go out and check the markings.

Ordered the spike part of the defense:
Screenshot (319)
 line:

Ordered the construction of the first few sections of the salt water pipe:
Screenshot (320)

- Return back, advance time, and finish the projects.
- Go out and check on the projects.

Finished the defense line:
Screenshot (321)

Finished the first few sections of the salt water pipe:
Screenshot (322)

- Return back and order the construction of the last salt water pipe segments.
- Go out and check on the projects.

Ordered the last part of the salt water pipe construction:
Screenshot (323)

- Return back and finish the projects.
- Start the construction of the salt water pump itself.

Ordered the salt water pump construction (and finished the last segment of the pipes):
Screenshot (324)

- Return back and finish the project.
- Go out and check the results.

Finished the salt water pump:
Screenshot (325)

#### Additional context

The new delete field operation works, as far as I can see, and I don't think it should be unsafe when used with fields that don't have effects beyond being visible. Comments from competent people would be desirable, however.

### CODE REVIEW

**Comment by ANickelN** [comment]
Does this work in showing the player a "preview" of what will change if the order is given?  I find frequently that base camp changes overlap with items or constructions I've placed.

**Comment by PatrikLundell** [comment]
No. As the images show, it only shows which tiles are affected by the construction mission, not what the mission constructs. Also, the indications are generated upon giving the construction order, at which time you have committed resources to the construction.
It does, however, show you where not to store stuff and park vehicles (and remove things if located there).

Displaying what to place where would require some illusionary version of everything terrain, furniture, and combination thereof that can be constructed, plus logic to tie the illusionary version to the real one and handle the cases where the illusionary version doesn't exist (because nobody has defined).

**Comment by IdleSol** [comment]
It's probably stupid.

What happens if I overlay the same field with ````"display_field": false,```? And then delete it.

**Comment by Procyonae** [comment]
This is great and I don't see any need for it to show a preview of the terrains and furniture

> Displaying what to place where would require some illusionary version of everything terrain, furniture, and combination thereof that can be constructed, plus logic to tie the illusionary version to the real one and handle the cases where the illusionary version doesn't exist (because nobody has defined it).

That said I do disagree that it'd need any of this, it could just use the existing sprites/symbols with some way to differentiate them like the way that seeing down z levels does, for sprites it could give them a blue hue and cut them up in lines to make it look like a hologram for example and for symbols the background could cycle light blue and back for instance.
Obviously I'm not suggesting this would be remotely easy to implement code wise though and would be very extra, but would be cool for say nest constructions in the future where you actually get to decide where they place and hence seeing what's actually being placed would be more relevant, even then just a different field for walls would probably do though.

**Comment by PatrikLundell** [comment]
You can't define the same thing twice (or if you can, only one of them would be active, probably "randomly", as that's what's happened in the past with duplicate definitions that weren't detected as duplicates).

It might be possible to hide it by defining a hiding field that has a blank as its symbol and has a higher priority, but then it would remain and clutter the memory (and hide any other future constructions in that space, such as e.g. the furnishing of a building just constructed). Much better to find out how to refresh the display of the fields.

I have no idea how to do any "graphic" display stuff, so illusory versions of things would have to be done by someone else. They'd have to be readily distinguishable from the real thing even in character mode.

File: data/json/field_type.json [hunk 1]
```diff
@@ -1985,5 +1985,13 @@
     "display_items": false,
     "display_field": true,
     "looks_like": "fd_tear_gas"
+  },
+  {
+    "id": "fd_construction_site",
+    "type": "field_type",
+    "intensity_levels": [ { "name": "Construction site", "sym": ".", "dangerous": false } ],
+    "priority": 1,
+    "display_field": true,
```
Comment (Maleclypse) [inline data/json/field_type.json:1994]:
```suggestion
    "display_field": true
```

File: data/json/field_type.json [hunk 2]
```diff
@@ -1985,5 +1985,13 @@
     "display_items": false,
     "display_field": true,
     "looks_like": "fd_tear_gas"
+  },
+  {
+    "id": "fd_construction_site",
+    "type": "field_type",
+    "intensity_levels": [ { "name": "Construction site", "sym": ".", "dangerous": false } ],
+    "priority": 1,
+    "display_field": true
```
Comment (Maleclypse) [inline data/json/field_type.json:1994]:
```suggestion
    "display_field": true,
```

**Comment by Maleclypse** [comment]
This is pretty cool!

**Comment by sparr** [comment]
Gonna steal this for another ui/visual-only field idea. Thanks for laying the groundwork!

**Pull request merged**